service/worker/sysworkflow/history_blob_iterator.go

```diff
@@ -0,0 +1,218 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package sysworkflow
+
+import (
+	"errors"
+	"strconv"
+	"time"
+
+	"github.com/uber-common/bark"
+	"github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/persistence"
+)
+
+/**
+IMPORTANT: Under the assumption that history is immutable, the following iterator will deterministically return identical
+blobs from Next regardless of the current cluster and regardless of the number of times this iteration occurs. This property
+makes any concurrent uploads of history safe.
+*/
+
+type (
+	// HistoryBlobIterator is used to get history blobs
+	HistoryBlobIterator interface {
+		Next() (*HistoryBlob, error)
+		HasNext() bool
+	}
+
+	historyBlobIterator struct {
+		logger        bark.Logger
+		metricsClient metrics.Client
+
+		// the following defines the state of the iterator
+		blobPageToken        int
+		persistencePageToken []byte
+		finishedIteration    bool
+
+		// the following are only used to read history and dynamic config
+		historyManager       persistence.HistoryManager
+		historyV2Manager     persistence.HistoryV2Manager
```

Comment:
Can we have 2 separate implementations of historyBlobIterator?  This will make things easier to remove any code related to historyV1Manager.

Reply:
We can. There were a couple reasons I did not do this in the the first version let me know what you think:

- The complexity of handling events_v1 and events_v2 is all isolated to one method in this module, so the transition to events_v2 should not be hard. 
- If we want to have two separate iterators the code is going to be more complex, we need to factor common code out into a base class and have two subclasses each of which basically only reads history and then call into the base class.